src/createServer.js

```diff
@@ -1,11 +1,280 @@
 'use strict';
 
-// const express = require('express');
+const express = require('express');
+const cors = require('cors');
+const uuid = require('uuid').v4;
 
 function createServer() {
-  // Use express to create a server
-  // Add a routes to the server
-  // Return the server (express app)
+  const collections = {
+    users: [],
+    expenses: [],
+  };
+
+  const app = express();
+
+  app.use(cors());
+
+  app.get('/users', (req, res) => {
+    res.send(collections.users);
+  });
+
+  app.get('users/:userId', (req, res) => {
+    const { userId } = req.params;
+    const foundUser = collections.users.find(user => user.id === userId);
+
+    if (!userId) {
+      res.sendStatus(400);
+
+      return;
+    }
+
+    if (!foundUser) {
+      res.sendStatus(404);
+
+      return;
+    }
+
+    res.sendStatus(200);
+    res.send(foundUser);
+  });
+
+  app.post('/users', express.json(), (req, res) => {
+    const { name } = req.body;
+
+    if (!name || Object.keys(req.body).length > 1) {
+      res.sendStatus(422);
+
+      return;
+    }
+
+    const newUser = {
+      id: uuid(),
+      name: name,
+    };
+
+    collections.users.push(newUser);
+
+    res.statusCode = 201;
+    res.send(newUser);
+  });
+
+  app.delete('users/:userId', (req, res) => {
+    const { userId } = req.params;
+    const foundUser = collections.users.find(user => user.id === userId);
+    const filteredUsers = collections.users.filter(user => user.id !== userId);
+
+    if (!foundUser) {
+      res.sendStatus(404);
+
+      return;
+    }
+
+    collections.users = filteredUsers;
+    res.statusCode = 204;
+    res.send(foundUser);
+  });
+
+  app.patch('users/:userId', express.json(), (req, res) => {
+    const { userId } = req.params;
+    const foundUser = collections.users.find(user => user.id === userId);
+
+    if (!userId) {
+      res.sendStatus(400);
+
+      return;
+    }
+
+    if (!foundUser) {
+      res.sendStatus(404);
+
+      return;
+    }
+
+    const { name } = req.body;
+
+    if (typeof name !== 'string') {
+      res.sendStatus(422);
+
+      return;
+    }
+
+    Object.assign(foundUser, { name });
+    res.statusCode = 200;
+    res.send(foundUser);
+  });
+
+  app.get('/expenses', (res, req) => {
+    res.send(collections.expenses);
+  });
+
+  app.get('/expenses/:purchaseId', (res, req) => {
+    const { purchaseId } = req.params;
+    const foundPurchase = collections.expenses
+      .find(purchase => purchase.id === purchaseId);
+
+    if (!purchaseId) {
+      res.sendStatus(400);
+
+      return;
+    }
+
+    if (!foundPurchase) {
+      res.sendStatus(404);
+
+      return;
+    }
+
+    res.sendStatus(200);
+    res.send(foundPurchase);
+  });
+
+  app.post('expenses/:purchaseId', express.json(), (res, req) => {
```

Comment:
```suggestion
  app.post('expenses/:purchaseId', express.json(), (req, res) => {
```
the first argument is request, the second one is response. Fix it everywhere, that's probably why tests fail